connexion/security.py

```diff
@@ -22,229 +55,269 @@
+        :param security_definition_key: The key which contains the value for the function name to resolve.
+        :param environ_key: The name of the environment variable that can be used alternatively for the function name.
+        :param default: The default to use in case the function cannot be found based on the security_definition_key or the environ_key
         """
+        # TODO: Turn this function's arguments into attributes instead?
         func = security_definition.get(security_definition_key) or os.environ.get(
             environ_key
         )
         if func:
             return get_function_from_name(func)
         return default
 
-    def get_tokeninfo_func(self, security_definition: dict) -> t.Optional[t.Callable]:
+    def _generic_check(self, func, exception_msg):
+        (
+            need_to_add_context,
+            need_to_add_required_scopes,
+        ) = self._need_to_add_context_or_scopes(func)
+
+        async def wrapper(request, *args, required_scopes=None):
+            kwargs = {}
+            if need_to_add_context:
+                kwargs[self.context_kw] = request.context
+            if need_to_add_required_scopes:
+                kwargs[self.required_scopes_kw] = required_scopes
+            token_info = func(*args, **kwargs)
+            while asyncio.iscoroutine(token_info):
+                token_info = await token_info
+            if token_info is self.no_value:
+                return self.no_value
+            if token_info is None:
+                raise OAuthResponseProblem(detail=exception_msg)
+            return token_info
+
+        return wrapper
+
+    # TODO: Convert to mixin?
+    @staticmethod
+    def get_auth_header_value(request):
         """
-        :type security_definition: dict
+        Called inside security wrapper functions
 
-        >>> get_tokeninfo_url({'x-tokenInfoFunc': 'foo.bar'})
-        '<function foo.bar>'
+        Return Authorization type and value if any.
+        If not Authorization, return (None, None)
+        Raise OAuthProblem for invalid Authorization header
         """
-        token_info_func = self._get_function(
-            security_definition, "x-tokenInfoFunc", "TOKENINFO_FUNC"
-        )
-        if token_info_func:
-            return token_info_func
+        authorization = request.headers.get("Authorization")
+        if not authorization:
+            return None, None
 
-        token_info_url = security_definition.get("x-tokenInfoUrl") or os.environ.get(
-            "TOKENINFO_URL"
-        )
-        if token_info_url:
-            return self.get_token_info_remote(token_info_url)
+        try:
+            auth_type, value = authorization.split(None, 1)
+        except ValueError:
+            raise OAuthProblem(detail="Invalid authorization header")
+        return auth_type.lower(), value
 
-        return None
+    def _need_to_add_context_or_scopes(self, func):
```

Comment:
For the context, it would make sense to keep only one way of accessing it instead of 2 ways to do the same thing.

Not sure about the required scopes:
in the past, only oauth and oidc were allowed to use scopes:
> For other security scheme types, the array MUST be empty.

but in OpenAPI 3.1, all security schemes can now use the scopes:
> For other security scheme types, the array MAY contain a list of role names which are required for the execution, but are not otherwise defined or exchanged in-band.

So it would depend on the signature of the security function again whether scopes need to be passed.